cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSQLEngineAdapter.java

```diff
@@ -17,112 +17,216 @@
+import io.cdap.cdap.api.metrics.Metrics;
+import io.cdap.cdap.etl.api.StageMetrics;
 import io.cdap.cdap.etl.api.engine.sql.SQLEngine;
-import io.cdap.cdap.etl.api.engine.sql.SQLOperationResult;
+import io.cdap.cdap.etl.api.engine.sql.SQLEngineException;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLDataset;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLPullDataset;
+import io.cdap.cdap.etl.api.engine.sql.dataset.SQLPushDataset;
 import io.cdap.cdap.etl.api.engine.sql.request.SQLJoinRequest;
 import io.cdap.cdap.etl.api.engine.sql.request.SQLPullRequest;
 import io.cdap.cdap.etl.api.engine.sql.request.SQLPushRequest;
 import io.cdap.cdap.etl.api.join.JoinDefinition;
+import io.cdap.cdap.etl.api.join.JoinStage;
+import io.cdap.cdap.etl.common.Constants;
+import io.cdap.cdap.etl.common.DefaultStageMetrics;
+import io.cdap.cdap.etl.common.StageStatisticsCollector;
 import io.cdap.cdap.etl.engine.SQLEngineJob;
+import io.cdap.cdap.etl.engine.SQLEngineJobKey;
 import io.cdap.cdap.etl.engine.SQLEngineJobType;
 import io.cdap.cdap.etl.spark.SparkCollection;
 import io.cdap.cdap.etl.spark.function.TransformFromPairFunction;
 import io.cdap.cdap.etl.spark.function.TransformToPairFunction;
 import org.apache.spark.api.java.JavaPairRDD;
 import org.apache.spark.api.java.JavaRDD;
 import org.apache.spark.api.java.JavaSparkContext;
-import org.apache.spark.api.java.function.FlatMapFunction;
-import org.apache.spark.api.java.function.PairFlatMapFunction;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-
-import static java.lang.Thread.currentThread;
+import javax.annotation.Nullable;
 
 /**
  * Adapter used to orchestrate interaction between the Pipeline Runner and the SQL Engine.
- * <p>
- * TODO: Add all operations supported by the SQL Engine.
+ *
+ * @param <T> type for records supported by this BatchSQLEngineAdapter.
  */
-public class BatchSQLEngineAdapter {
+public class BatchSQLEngineAdapter<T> {
   private static final Logger LOG = LoggerFactory.getLogger(BatchSQLEngineAdapter.class);
 
-  SQLEngine<?, ?, ?, ?> sqlEngine;
-  ExecutorService executorService = Executors.newFixedThreadPool(10);
-  PairFlatMapFunction<?, ?, ?> toPairFunction;
-  FlatMapFunction<?, ?> fromPairFunction;
-  Map<String, SQLEngineJob<?>> jobs;
+  private final SQLEngine<?, ?, ?, ?> sqlEngine;
+  private final Metrics metrics;
+  private final Map<String, StageStatisticsCollector> statsCollectors;
+  private final ExecutorService executorService = Executors.newCachedThreadPool();
```

Comment:
should specify the ThreadFactory with a descriptive thread name (see other places where we use Execturos.new...)

Reply:
Oh yes, and usually you want to make you class closable and close this guy on shutdown to prevent non-daemon threads to hang around preventing your app closure.

Reply:
Added.

Reply:
Added this.